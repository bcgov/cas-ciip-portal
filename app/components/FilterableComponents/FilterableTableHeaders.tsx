import React, {useState, useEffect} from 'react';
import {Button, Form} from 'react-bootstrap';
import {useRouter} from 'next/router';
import {ISearchProps} from 'components/Search/SearchProps';
import {getUserFriendlyStatusLabel} from 'lib/text-transforms';

const NONE_VALUES = [null, undefined];

const FilterableTableHeaders: React.FunctionComponent<ISearchProps> = (
  props
) => {
  const router = useRouter();

  const [searchFilters, setSearchFilters] = useState({});

  const handleFilterChange = (value, column, parseMethod) => {
    setSearchFilters({
      ...searchFilters,
      [column]: parseMethod ? parseMethod(value) : value
    });
  };

  const applySearch = (searchData) => {
    const newQuery = {};
    props.searchOptions.forEach((option) => {
      const column = option.columnName;

      if (option.isSearchEnabled) {
        if (!NONE_VALUES.includes(searchData[column]))
          newQuery[column] = searchData[column];
        else delete newQuery[column];
      }
    });

    const queryString = JSON.stringify(newQuery);

    const url = {
      pathname: router.pathname,
      query: {
        ...router.query,
        relayVars: queryString
      }
    };

    router.replace(url, url, {shallow: true});
  };

  const clearForm = () => {
    setSearchFilters({});
    applySearch({});
  };

  // functional equivalent to componentDidMount()
  useEffect(() => {
    const parsedUrl = router.query.relayVars
      ? JSON.parse(String(router.query.relayVars))
      : {};
    Object.keys(parsedUrl).forEach((key) => {
      if (NONE_VALUES.includes(searchFilters[key]))
        searchFilters[key] = parsedUrl[key];
    });
  }, []);

  return (
    <tr>
      {props.searchOptions.map((option) => {
        const column = option.columnName;
        const key = option.columnName + option.title;
        const initialValue = searchFilters[column] ?? '';

        if (option.isSearchEnabled) {
          if (!option.searchOptionValues) {
            return (
              <td key={key}>
                <Form.Control
                  placeholder="Search"
                  name={column}
                  value={initialValue}
                  onChange={(evt) =>
                    handleFilterChange(evt.target.value, column, option.toUrl)
                  }
                />
              </td>
            );
          }

          return (
            <td key={key}>
              <Form.Control
                as="select"
                placeholder="Search"
                name={column}
                value={initialValue}
                onChange={(evt) =>
                  handleFilterChange(evt.target.value, column, option.toUrl)
                }
              >
                <option key={column + '-placeholder'} value="">
                  ...
                </option>
                {option.searchOptionValues.map((kvp) => (
                  <option key={column + '-' + kvp.display} value={kvp.value}>
                    {getUserFriendlyStatusLabel(kvp.display)}
                  </option>
                ))}
              </Form.Control>
            </td>
          );
        }
        if (option.removeSearchHeader) return;
        return <td key={key} />;
      })}
      <td className="flex">
        <Button variant="outline-secondary" onClick={clearForm}>
          Clear
        </Button>
        <Button
          style={{marginLeft: '5px'}}
          variant="primary"
          onClick={() => applySearch(searchFilters)}
        >
          Apply
        </Button>
      </td>
      <style jsx>{`
        .flex {
          display: flex;
        }
      `}</style>
    </tr>
  );
};

export default FilterableTableHeaders;
