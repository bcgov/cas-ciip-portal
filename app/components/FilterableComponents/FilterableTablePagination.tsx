import React from 'react';
import {useRouter} from 'next/router';
import {Pagination} from 'react-bootstrap';

interface Props {
  totalCount: number;
  maxResultsPerPage: number;
}

export const FilterableTablePaginationComponent: React.FunctionComponent<Props> = (
  props
) => {
  const {totalCount, maxResultsPerPage} = props;
  const router = useRouter();
  const maxPages = Math.ceil(totalCount / maxResultsPerPage);

  // Set the activePage by query string value. If no value exists, the activePage is the first page
  let activePage = 1;
  if (router?.query?.pageVars)
    activePage =
      JSON.parse(router?.query?.pageVars?.toString())?.activePage || 1;

  const items = [];

  const paginate = (pageNumber: number) => {
    const paginationObject = {
      offset: (pageNumber - 1) * maxResultsPerPage,
      activePage: pageNumber
    };
    const url = {
      pathname: router.pathname,
      query: {
        ...router.query,
        pageVars: JSON.stringify(paginationObject)
      }
    };
    router.push(url, url, {shallow: true});
  };

  // Determines what page numbers to render based on the location of the activePage
  let startPage;
  let endPage;
  if (maxPages <= 9) {
    startPage = 1;
    endPage = maxPages;
  } else if (maxPages - activePage <= 4) {
    startPage = maxPages - 8;
    endPage = maxPages;
  } else if (activePage <= 5) {
    startPage = 1;
    endPage = 9;
  } else {
    startPage = activePage - 4;
    endPage = activePage + 4;
  }

  // Populate the Pagination items with page numbers & functionality
  for (let pageNumber = startPage; pageNumber <= endPage; pageNumber++) {
    items.push(
      <Pagination.Item
        key={pageNumber}
        active={pageNumber === activePage}
        onClick={() => paginate(pageNumber)}
      >
        {pageNumber}
      </Pagination.Item>
    );
  }
  if (maxPages > 1)
    return (
      <Pagination>
        <Pagination.First onClick={() => paginate(1)} />
        <Pagination.Prev
          onClick={() => (activePage > 1 ? paginate(activePage - 1) : null)}
        />
        {startPage !== 1 && <Pagination.Ellipsis />}
        {items}
        {endPage !== maxPages && <Pagination.Ellipsis />}
        <Pagination.Next
          onClick={() =>
            activePage !== maxPages ? paginate(activePage + 1) : null
          }
        />
        <Pagination.Last onClick={() => paginate(maxPages)} />
      </Pagination>
    );
  return null;
};

export default FilterableTablePaginationComponent;
