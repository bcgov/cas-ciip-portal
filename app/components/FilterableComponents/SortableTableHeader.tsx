import React, {useState} from 'react';
import {FontAwesomeIcon} from '@fortawesome/react-fontawesome';
import {
  faSort,
  faCaretUp,
  faCaretDown
} from '@fortawesome/free-solid-svg-icons';
import {useRouter} from 'next/router';

interface Props {
  headerVariables: {
    columnName: string;
    displayName: string;
    sortable: boolean;
  };
}

const SORT_DIRECTION = ['ASC', 'DESC'];
const SORT_ICONS = [faCaretDown, faCaretUp];

const SortableTableHeader: React.FunctionComponent<Props> = ({
  headerVariables
}) => {
  const router = useRouter();
  const sortDirectionIndex = SORT_DIRECTION.indexOf(
    router.query.direction?.toString()
  );
  const [currentSortDirection, setCurrentSortDirection] = useState(
    Math.max(sortDirectionIndex, 0)
  );

  const getOrderbyString = (orderColumnName, sortDirection) => {
    return orderColumnName.toUpperCase() + '_' + SORT_DIRECTION[sortDirection];
  };

  const triggerSort = (sortColumnName) => {
    //Cycle
    const sortDirection = (currentSortDirection + 1) % 2;
    setCurrentSortDirection(sortDirection);

    const url = {
      pathname: router.pathname,
      query: {
        ...router.query,
        order_by: getOrderbyString(sortColumnName, sortDirection)
      }
    };

    router.replace(url, url, {shallow: true});
  };

  const {columnName, displayName, sortable} = headerVariables;
  return (
    <th onClick={() => triggerSort(columnName)}>
      <span>{displayName}</span>
      {sortable && (
        <span style={{height: '100%', position: 'absolute', right: '0.75em'}}>
          <FontAwesomeIcon
            color="white"
            icon={
              router.query.order_by ===
              getOrderbyString(columnName, currentSortDirection)
                ? SORT_ICONS[currentSortDirection]
                : faSort
            }
          />
        </span>
      )}
      <style>{`
        .table thead th {
          background: #003366;
          color: white;
          position: relative;
          padding-right: calc(1em + 10px);
          cursor: pointer;
        }
      `}</style>
    </th>
  );
};

export default SortableTableHeader;
