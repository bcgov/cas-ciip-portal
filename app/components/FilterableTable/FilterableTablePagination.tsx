import React from "react";
import { Pagination, Dropdown } from "react-bootstrap";

interface Props {
  /**
   * The total number of items in all of the pages
   */
  totalCount: number;
  /**
   * Defaults to DEFAULT_PAGE_SIZE.
   */
  pageSize?: number;
  /**
   * The number of items to skip to get to the current page. Defaults to 0
   */
  offset?: number;
  onOffsetChange: (offset: number) => void;
  onPageSizeChange: (pageSize: number) => void;
}

export const DEFAULT_PAGE_SIZE = 20;

const PAGE_SIZE_OPTIONS = [20, 50, 100];

const FilterableTablePagination: React.FunctionComponent<Props> = ({
  totalCount,
  pageSize = DEFAULT_PAGE_SIZE,
  offset = 0,
  onOffsetChange,
  onPageSizeChange,
}) => {
  const activePage = Math.floor(offset / pageSize + 1) || 1;
  const maxPages = Math.ceil(totalCount / pageSize);

  const items = [];

  // Determines what page numbers to render based on the location of the activePage
  let startPage;
  let endPage;
  if (maxPages <= 9) {
    startPage = 1;
    endPage = maxPages;
  } else if (maxPages - activePage <= 4) {
    startPage = maxPages - 8;
    endPage = maxPages;
  } else if (activePage <= 5) {
    startPage = 1;
    endPage = 9;
  } else {
    startPage = activePage - 4;
    endPage = activePage + 4;
  }

  const handlePageChange = (pageNumber: number) => {
    onOffsetChange((pageNumber - 1) * pageSize);
  };

  // Populate the Pagination items with page numbers & functionality
  for (let pageNumber = startPage; pageNumber <= endPage; pageNumber++) {
    items.push(
      <Pagination.Item
        key={pageNumber}
        active={pageNumber === activePage}
        onClick={() => handlePageChange(pageNumber)}
      >
        {pageNumber}
      </Pagination.Item>
    );
  }
  return (
    <>
      <div className="pagination">
        <Dropdown>
          <Dropdown.Toggle variant="outline-dark" id="dropdown-basic">
            Items Per Page: <strong>{pageSize}</strong>
          </Dropdown.Toggle>

          <Dropdown.Menu>
            {PAGE_SIZE_OPTIONS.map((option) => (
              <Dropdown.Item
                key={option}
                onClick={() => onPageSizeChange(option)}
              >
                {option}
              </Dropdown.Item>
            ))}
          </Dropdown.Menu>
        </Dropdown>
        {maxPages > 1 && (
          <Pagination style={{ marginLeft: "15px" }}>
            <Pagination.First onClick={() => handlePageChange(1)}>
              &lt;&lt; First Page
            </Pagination.First>
            <Pagination.Prev
              onClick={() =>
                activePage > 1 ? handlePageChange(activePage - 1) : null
              }
            />
            {startPage !== 1 && <Pagination.Ellipsis />}
            {items}
            {endPage !== maxPages && <Pagination.Ellipsis />}
            <Pagination.Next
              onClick={() =>
                activePage !== maxPages
                  ? handlePageChange(activePage + 1)
                  : null
              }
            />
            <Pagination.Last onClick={() => handlePageChange(maxPages)}>
              Last Page &gt;&gt;
            </Pagination.Last>
          </Pagination>
        )}
      </div>
      <style jsx>{`
        .pagination {
          display: flex;
        }
        :global(.pagination a.page-link) {
          color: #036;
        }
        :global(.pagination .page-item.active .page-link) {
          background-color: #036;
          border-color: #036;
        }
      `}</style>
    </>
  );
};

export default FilterableTablePagination;
