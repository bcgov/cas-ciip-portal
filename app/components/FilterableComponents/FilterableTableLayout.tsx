import React, {useMemo} from 'react';
import {Col, Table, Alert} from 'react-bootstrap';
import SortableTableHeader from './SortableTableHeader';
import FilterableTableHeaders from './FilterableTableHeaders';
import {ISearchProps, ISearchExtraFilter} from 'components/Search/SearchProps';
import {useRouter} from 'next/router';
import safeJsonParse from 'lib/safeJsonParse';
import {FilterArgs} from 'components/Search/ISearchOption';

interface Props extends ISearchProps {
  body: JSX.Element;
  isLoading?: boolean;
  extraFilters?: ISearchExtraFilter[];
}
export const FilterableTableLayoutComponent: React.FunctionComponent<Props> = ({
  searchOptions,
  body,
  isLoading,
  extraFilters = []
}) => {
  const router = useRouter();
  const relayVars = useMemo(
    () => safeJsonParse(router.query.relayVars as string),
    [router]
  );

  const applySearch = (searchData: FilterArgs) => {
    const newQuery = {
      // copy the vars from the query string, so that the args coming from extraFilters are not overriden
      ...relayVars,
      ...searchData
    };
    searchOptions.forEach((option) => {
      const column = option.columnName;

      if (option.isSearchEnabled) {
        newQuery[column] = searchData[column] ?? undefined;
      }
    });

    const queryString = JSON.stringify(newQuery);

    const url = {
      pathname: router.pathname,
      query: {
        ...router.query,
        relayVars: queryString,
        pageVars: JSON.stringify({offset: 0})
      }
    };

    router.push(url, url, {shallow: true});
  };

  const bodyLength =
    body.props.children?.length ?? body.props.children?.[0]?.length ?? 0;
  const noSearchResults =
    bodyLength === 0 ? (
      <Alert variant="secondary" id="no-search-results">
        No matching results to show.
      </Alert>
    ) : null;

  return (
    <>
      <div>
        {extraFilters.map((f) => {
          return (
            <f.Component
              key={f.argName}
              onChange={(value) =>
                applySearch({...relayVars, [f.argName]: value})
              }
              value={relayVars[f.argName]}
            />
          );
        })}
      </div>
      <Table
        striped
        bordered
        hover
        className={isLoading ? 'search-table loading' : 'search-table'}
      >
        <thead>
          <tr>
            {searchOptions.map((option) => (
              <SortableTableHeader
                key={option.title + '-sortHeader'}
                headerVariables={{
                  columnName: option.columnName,
                  displayName: option.title,
                  sortable: option.isSortEnabled
                }}
              />
            ))}
          </tr>
          <FilterableTableHeaders
            filterArgs={relayVars}
            searchOptions={searchOptions}
            onSubmit={applySearch}
          />
        </thead>
        {body}
      </Table>
      <Col md={{span: 6, offset: 3}}>{noSearchResults}</Col>
      <style jsx global>{`
        .search-table {
          text-align: center;
        }
        .master-select {
          background: #003366;
          color: white;
          min-width: 6em;
          position: relative;
        }
        .master-select label {
          margin-bottom: 0;
          padding-right: 24px;
          width: 100%;
          display: flex;
          flex-direction: column;
          align-items: center;
          justify-content: center;
          position: absolute;
          top: 0;
          right: 0;
          bottom: 0;
          left: 0;
        }
        .master-select input[type='checkbox'] {
          position: absolute;
          right: 10px;
          top: calc(50% - 6px);
        }
        .search-table.loading td {
          opacity: 0;
          pointer-events: none;
        }

        .search-table.loading tbody {
          position: relative;
        }
        .search-table.loading tbody::after {
          width: 2rem;
          height: 2rem;
          vertical-align: text-bottom;
          background-color: black;
          border-radius: 50%;
          opacity: 0;
          content: '';
          position: absolute;
          top: 50%;
          left: 50%;
          animation: spinner-grow 0.75s linear infinite;
        }
        #no-search-results {
          text-align: center;
        }
      `}</style>
    </>
  );
};

export default FilterableTableLayoutComponent;
