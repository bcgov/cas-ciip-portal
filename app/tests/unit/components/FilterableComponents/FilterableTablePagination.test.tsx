import React from 'react';
import {shallow} from 'enzyme';
import {FilterableTablePaginationComponent} from 'components/FilterableComponents/FilterableTablePagination';
import * as nextRouter from 'next/router';

nextRouter.useRouter = jest.fn();
nextRouter.useRouter.mockImplementation(() => ({
  route: '/',
  query: {
    pageVars: null
  }
}));

describe('PaginationBar', () => {
  it('should not render the Pagination component if Math.ceil(totalCount / maxResultsPerPage) < 2', async () => {
    const r = shallow(
      <FilterableTablePaginationComponent
        totalCount={5}
        maxResultsPerPage={10}
      />
    );
    expect(r.find('PageItem').exists()).toBe(false);
    expect(r).toMatchSnapshot();
  });
  it('should render the Pagination component if Math.ceil(totalCount / maxResultsPerPage) > 1', async () => {
    const r = shallow(
      <FilterableTablePaginationComponent
        totalCount={25}
        maxResultsPerPage={10}
      />
    );
    expect(r.find('PageItem').exists()).toBe(true);
    expect(r).toMatchSnapshot();
  });
  it('should create a number of pagination buttons equal to Math.ceil(totalCount / maxResultsPerPage)', async () => {
    const r = shallow(
      <FilterableTablePaginationComponent
        totalCount={25}
        maxResultsPerPage={10}
      />
    );
    expect(r.find('PageItem').at(0).text()).toBe('1');
    expect(r.find('PageItem').at(1).text()).toBe('2');
    expect(r.find('PageItem').at(2).text()).toBe('3');
    expect(r.find('PageItem').at(3).exists()).toBe(false);
  });
  it('should render a maximum of 9 pagination pages if Math.ceil(totalCount / maxResultsPerPage) > 9', async () => {
    const r = shallow(
      <FilterableTablePaginationComponent
        totalCount={130}
        maxResultsPerPage={10}
      />
    );
    expect(r.find('PageItem').at(8).text()).toBe('9');
    expect(r.find('PageItem').at(9).exists()).toBe(false);
    expect(r.find('Ellipsis').exists()).toBe(true);
    expect(r).toMatchSnapshot();
  });
});
