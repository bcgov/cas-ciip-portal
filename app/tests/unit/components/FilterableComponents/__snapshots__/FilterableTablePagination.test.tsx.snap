// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginationBar should not render the Pagination component if Math.ceil(totalCount / maxResultsPerPage) < 2 1`] = `""`;

exports[`PaginationBar should render a maximum of 9 pagination pages if Math.ceil(totalCount / maxResultsPerPage) > 9 1`] = `
<ForwardRef>
  <First
    onClick={[Function]}
  />
  <Prev
    onClick={[Function]}
  />
  <PageItem
    active={true}
    activeLabel="(current)"
    disabled={false}
    key="1"
    onClick={[Function]}
  >
    1
  </PageItem>
  <PageItem
    active={false}
    activeLabel="(current)"
    disabled={false}
    key="2"
    onClick={[Function]}
  >
    2
  </PageItem>
  <PageItem
    active={false}
    activeLabel="(current)"
    disabled={false}
    key="3"
    onClick={[Function]}
  >
    3
  </PageItem>
  <PageItem
    active={false}
    activeLabel="(current)"
    disabled={false}
    key="4"
    onClick={[Function]}
  >
    4
  </PageItem>
  <PageItem
    active={false}
    activeLabel="(current)"
    disabled={false}
    key="5"
    onClick={[Function]}
  >
    5
  </PageItem>
  <PageItem
    active={false}
    activeLabel="(current)"
    disabled={false}
    key="6"
    onClick={[Function]}
  >
    6
  </PageItem>
  <PageItem
    active={false}
    activeLabel="(current)"
    disabled={false}
    key="7"
    onClick={[Function]}
  >
    7
  </PageItem>
  <PageItem
    active={false}
    activeLabel="(current)"
    disabled={false}
    key="8"
    onClick={[Function]}
  >
    8
  </PageItem>
  <PageItem
    active={false}
    activeLabel="(current)"
    disabled={false}
    key="9"
    onClick={[Function]}
  >
    9
  </PageItem>
  <Ellipsis />
  <Next
    onClick={[Function]}
  />
  <Last
    onClick={[Function]}
  />
</ForwardRef>
`;

exports[`PaginationBar should render the Pagination component if Math.ceil(totalCount / maxResultsPerPage) > 1 1`] = `
<ForwardRef>
  <First
    onClick={[Function]}
  />
  <Prev
    onClick={[Function]}
  />
  <PageItem
    active={true}
    activeLabel="(current)"
    disabled={false}
    key="1"
    onClick={[Function]}
  >
    1
  </PageItem>
  <PageItem
    active={false}
    activeLabel="(current)"
    disabled={false}
    key="2"
    onClick={[Function]}
  >
    2
  </PageItem>
  <Next
    onClick={[Function]}
  />
  <Last
    onClick={[Function]}
  />
</ForwardRef>
`;

exports[`PaginationBar should still render if there is bad data in the querystring 1`] = `
<ForwardRef>
  <First
    onClick={[Function]}
  />
  <Prev
    onClick={[Function]}
  />
  <PageItem
    active={true}
    activeLabel="(current)"
    disabled={false}
    key="1"
    onClick={[Function]}
  >
    1
  </PageItem>
  <PageItem
    active={false}
    activeLabel="(current)"
    disabled={false}
    key="2"
    onClick={[Function]}
  >
    2
  </PageItem>
  <PageItem
    active={false}
    activeLabel="(current)"
    disabled={false}
    key="3"
    onClick={[Function]}
  >
    3
  </PageItem>
  <PageItem
    active={false}
    activeLabel="(current)"
    disabled={false}
    key="4"
    onClick={[Function]}
  >
    4
  </PageItem>
  <PageItem
    active={false}
    activeLabel="(current)"
    disabled={false}
    key="5"
    onClick={[Function]}
  >
    5
  </PageItem>
  <PageItem
    active={false}
    activeLabel="(current)"
    disabled={false}
    key="6"
    onClick={[Function]}
  >
    6
  </PageItem>
  <PageItem
    active={false}
    activeLabel="(current)"
    disabled={false}
    key="7"
    onClick={[Function]}
  >
    7
  </PageItem>
  <Next
    onClick={[Function]}
  />
  <Last
    onClick={[Function]}
  />
</ForwardRef>
`;
